File: python/tests/attachments_test.py
"""Attachments (S-AT) acceptance scenarios, including the catching halves of the
G-S3 case (download authorization derives from channel membership).
"""

from __future__ import annotations

from harness.fixture import Fixture
from tests.helpers import client_for, seed_channel, seed_user


def _upload(
    fixture: Fixture, user_id: str, channel_id: str, content: bytes, filename: str = "f.bin"
):
    return client_for(fixture, user_id).post_multipart(
        f"/channels/{channel_id}/attachments",
        {"file": (filename, content, "application/octet-stream")},
    )


def test_s_at_01_member_uploads(fixture: Fixture) -> None:
    owner = seed_user(fixture, "atowner")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    content = b"x" * (10 * 1024)
    response = _upload(fixture, owner.id, public_id, content, "doc.bin")
    response.expect(201)
    body = response.json()
    assert body["sizeBytes"] == len(content)
    assert body["filename"] == "doc.bin"
    # Bytes landed in MinIO under the storage key (channel/id).
    assert fixture.s3.object_bytes(f"{public_id}/{body['id']}") == content


def test_s_at_02_non_member_upload(fixture: Fixture) -> None:
    owner = seed_user(fixture, "atowner")
    outsider = seed_user(fixture, "atout")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    private_id = seed_channel(fixture, owner, "priv", private=True)
    _upload(fixture, outsider.id, public_id, b"data").expect(403)
    _upload(fixture, outsider.id, private_id, b"data").expect(404)


def test_s_at_03_size_limits(fixture: Fixture) -> None:
    owner = seed_user(fixture, "atowner")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    _upload(fixture, owner.id, public_id, b"x" * (1024 * 1024 + 1)).expect(413).expect_code(
        "attachment:too_large"
    )
    _upload(fixture, owner.id, public_id, b"").expect(422).expect_code("attachment:empty")


def test_s_at_04_reference_attachment_in_message(fixture: Fixture) -> None:
    owner = seed_user(fixture, "atowner")
    other = seed_user(fixture, "atother")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    other_channel = seed_channel(fixture, owner, "pub2", private=False)
    client_for(fixture, other.id).post(f"/channels/{public_id}/join").expect(201)

    mine = _upload(fixture, owner.id, public_id, b"mine").json()["id"]
    response = client_for(fixture, owner.id).post(
        f"/channels/{public_id}/messages", {"text": "see attached", "attachmentIds": [mine]}
    )
    response.expect(201)
    assert response.json()["attachmentIds"] == [mine]

    # Referencing another user's attachment → 422.
    theirs = _upload(fixture, other.id, public_id, b"theirs").json()["id"]
    client_for(fixture, owner.id).post(
        f"/channels/{public_id}/messages", {"text": "x", "attachmentIds": [theirs]}
    ).expect(422).expect_code("message:attachment:invalid")

    # Referencing an attachment from another channel → 422.
    elsewhere = _upload(fixture, owner.id, other_channel, b"elsewhere").json()["id"]
    client_for(fixture, owner.id).post(
        f"/channels/{public_id}/messages", {"text": "x", "attachmentIds": [elsewhere]}
    ).expect(422).expect_code("message:attachment:invalid")


def test_s_at_05_member_downloads(fixture: Fixture) -> None:
    owner = seed_user(fixture, "atowner")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    content = b"download-me"
    attachment_id = _upload(fixture, owner.id, public_id, content, "report.pdf").json()["id"]

    response = client_for(fixture, owner.id).get(f"/attachments/{attachment_id}")
    response.expect(200)
    assert response.raw == content
    assert "report.pdf" in response.headers["content-disposition"]


def test_s_at_06_s3_private_non_member_download_404(fixture: Fixture) -> None:
    """[G-S3] non-member downloads private-channel attachment by id → 404, zero
    bytes; body identical to unknown-id 404."""
    owner = seed_user(fixture, "atowner")
    outsider = seed_user(fixture, "atout")
    private_id = seed_channel(fixture, owner, "priv", private=True)
    attachment_id = _upload(fixture, owner.id, private_id, b"secret-bytes").json()["id"]

    forbidden = client_for(fixture, outsider.id).get(f"/attachments/{attachment_id}")
    forbidden.expect(404).expect_code("attachment:not_found")
    assert b"secret-bytes" not in forbidden.raw
    unknown = client_for(fixture, outsider.id).get("/attachments/does-not-exist")
    unknown.expect(404)
    assert forbidden.raw == unknown.raw


def test_s_at_07_s3_public_non_member_download_403(fixture: Fixture) -> None:
    """[G-S3] non-member downloads public-channel attachment → 403, zero bytes."""
    owner = seed_user(fixture, "atowner")
    outsider = seed_user(fixture, "atout")
    public_id = seed_channel(fixture, owner, "pub", private=False)
    attachment_id = _upload(fixture, owner.id, public_id, b"public-bytes").json()["id"]

    response = client_for(fixture, outsider.id).get(f"/attachments/{attachment_id}")
    response.expect(403).expect_code("channel:membership_required")
    assert b"public-bytes" not in response.raw

File: go/tests/attachments_test.go
package relaytest

import (
	"bytes"
	"mime/multipart"
	"net/http"
	"testing"

	"github.com/drim-dev/verifying-agent-code/go/src/relay/app"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/domain"
)

func TestS_AT_01_upload_stores_bytes(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	content := bytes.Repeat([]byte("x"), 10*1024)
	id := uploadAttachment(t, ada, ch, "notes.txt", content).expect(http.StatusCreated).attachmentID(t)

	ctx, cancel := bgCtx()
	defer cancel()
	key := ch + "/" + id
	got, err := fixture.S3.ObjectBytes(ctx, key)
	if err != nil {
		t.Fatalf("object bytes: %v", err)
	}
	if !bytes.Equal(got, content) {
		t.Fatalf("stored bytes differ from upload (%d vs %d)", len(got), len(content))
	}
	if n := dbCount(t, "attachments", "id = $1 AND channel_id = $2 AND uploader_id = $3", id, ch, ada.ID); n != 1 {
		t.Fatalf("expected one attachment metadata row, got %d", n)
	}
}

func TestS_AT_02_upload_visibility(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	pub := seedChannel(t, ada, "public", false)
	priv := seedChannel(t, ada, "secret", true)
	uploadAttachment(t, bob, pub, "f.txt", []byte("hi")).
		expect(http.StatusForbidden).expectCode("channel:membership_required")
	uploadAttachment(t, bob, priv, "f.txt", []byte("hi")).
		expect(http.StatusNotFound).expectCode("channel:not_found")
}

func TestS_AT_03_upload_size_limits(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	tooBig := bytes.Repeat([]byte("x"), (1<<20)+1)
	uploadAttachment(t, ada, ch, "big.bin", tooBig).
		expect(http.StatusRequestEntityTooLarge).expectCode("attachment:too_large")
	uploadAttachment(t, ada, ch, "empty.txt", []byte{}).
		expect(http.StatusUnprocessableEntity).expectCode("attachment:empty")
}

func TestS_AT_04_message_attachment_reference(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	ch := seedChannel(t, ada, "team", false)
	other := seedChannel(t, ada, "other", false)
	seedJoin(t, bob, ch)

	mine := uploadAttachment(t, ada, ch, "mine.txt", []byte("mine")).expect(http.StatusCreated).attachmentID(t)
	bobs := uploadAttachment(t, bob, ch, "bobs.txt", []byte("bobs")).expect(http.StatusCreated).attachmentID(t)
	elsewhere := uploadAttachment(t, ada, other, "x.txt", []byte("x")).expect(http.StatusCreated).attachmentID(t)

	var msg struct {
		ID string `json:"id"`
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "see attached", "attachmentIds": []string{mine}}).
		expect(http.StatusCreated).decode(&msg)
	if n := dbCount(t, "attachments", "id = $1 AND message_id = $2", mine, msg.ID); n != 1 {
		t.Fatalf("attachment should be bound to the message, got %d", n)
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "x", "attachmentIds": []string{bobs}}).
		expect(http.StatusUnprocessableEntity).expectCode("message:attachment:invalid")
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "x", "attachmentIds": []string{elsewhere}}).
		expect(http.StatusUnprocessableEntity).expectCode("message:attachment:invalid")
}

func TestS_AT_05_download_returns_bytes(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	content := []byte("hello attachment")
	id := uploadAttachment(t, ada, ch, "greeting.txt", content).expect(http.StatusCreated).attachmentID(t)

	resp := newClient(t, ada.ID).get("/attachments/" + id)
	resp.expect(http.StatusOK)
	if !bytes.Equal(resp.body, content) {
		t.Fatalf("downloaded bytes differ from upload")
	}
	if cd := resp.header.Get("Content-Disposition"); cd == "" || !bytes.Contains([]byte(cd), []byte("greeting.txt")) {
		t.Fatalf("expected filename in Content-Disposition, got %q", cd)
	}
}

// ---- G-S3: download authorization is channel membership, never id possession ----

func TestS_AT_06_private_download_hidden(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	priv := seedChannel(t, ada, "secret", true)
	id := uploadAttachment(t, ada, priv, "secret.bin", []byte("classified")).expect(http.StatusCreated).attachmentID(t)
	assertPrivateAttachmentHidden(t, fixture.BaseURL(), bob, id)
}

func TestS_AT_06_naive_id_possession_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	priv := seedChannel(t, ada, "secret", true)
	id := uploadAttachment(t, ada, priv, "secret.bin", []byte("classified")).expect(http.StatusCreated).attachmentID(t)
	naive := fixture.NaiveApp(func(d *app.Deps) { d.AttachmentAccess = naiveAttachmentAccess{store: fixture.Store} })
	defer naive.Close()
	expectCatchToFail(t, "G-S3", func(tb testing.TB) {
		assertPrivateAttachmentHidden(tb, naive.BaseURL(), bob, id)
	})
}

func assertPrivateAttachmentHidden(tb testing.TB, baseURL string, outsider domain.User, attachmentID string) {
	tb.Helper()
	visible := clientAt(tb, baseURL, outsider.ID).get("/attachments/" + attachmentID)
	visible.expect(http.StatusNotFound).expectCode("attachment:not_found")
	if bytes.Contains(visible.body, []byte("classified")) {
		tb.Fatalf("private attachment bytes leaked to a non-member")
	}
	unknown := clientAt(tb, baseURL, outsider.ID).get("/attachments/UNKNOWN0000000")
	if visible.rawBody() != unknown.rawBody() {
		tb.Fatalf("existence-hiding body differs:\n visible: %s\n unknown: %s", visible.rawBody(), unknown.rawBody())
	}
}

func TestS_AT_07_public_download_requires_membership(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	pub := seedChannel(t, ada, "public", false)
	id := uploadAttachment(t, ada, pub, "shared.bin", []byte("shared")).expect(http.StatusCreated).attachmentID(t)
	assertPublicAttachmentForbidden(t, fixture.BaseURL(), bob, id)
}

func TestS_AT_07_naive_public_id_possession_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	pub := seedChannel(t, ada, "public", false)
	id := uploadAttachment(t, ada, pub, "shared.bin", []byte("shared")).expect(http.StatusCreated).attachmentID(t)
	naive := fixture.NaiveApp(func(d *app.Deps) { d.AttachmentAccess = naiveAttachmentAccess{store: fixture.Store} })
	defer naive.Close()
	expectCatchToFail(t, "G-S3", func(tb testing.TB) {
		assertPublicAttachmentForbidden(tb, naive.BaseURL(), bob, id)
	})
}

func assertPublicAttachmentForbidden(tb testing.TB, baseURL string, outsider domain.User, attachmentID string) {
	tb.Helper()
	resp := clientAt(tb, baseURL, outsider.ID).get("/attachments/" + attachmentID)
	resp.expect(http.StatusForbidden).expectCode("channel:membership_required")
	if bytes.Contains(resp.body, []byte("shared")) {
		tb.Fatalf("public attachment bytes leaked to a non-member")
	}
}

// uploadAttachment posts a multipart file as the given user and returns the response.
func uploadAttachment(t *testing.T, by domain.User, channelID, filename string, content []byte) *response {
	t.Helper()
	var buf bytes.Buffer
	mw := multipart.NewWriter(&buf)
	fw, err := mw.CreateFormFile("file", filename)
	if err != nil {
		t.Fatalf("create form file: %v", err)
	}
	if _, err := fw.Write(content); err != nil {
		t.Fatalf("write form file: %v", err)
	}
	if err := mw.Close(); err != nil {
		t.Fatalf("close writer: %v", err)
	}
	req, err := http.NewRequest(http.MethodPost, fixture.BaseURL()+"/channels/"+channelID+"/attachments", &buf)
	if err != nil {
		t.Fatalf("new request: %v", err)
	}
	req.Header.Set("Content-Type", mw.FormDataContentType())
	return newClient(t, by.ID).doRaw(req)
}

func (r *response) attachmentID(t *testing.T) string {
	t.Helper()
	var a struct {
		ID string `json:"id"`
	}
	r.decode(&a)
	return a.ID
}
